src/lib/proof-system/zkFunction.ts

```diff
@@ -0,0 +1,157 @@
+import { Snarky, initializeBindings } from '../../snarky.js';
+import { MlFieldArray, MlFieldConstArray } from '../ml/fields.js';
+import { withThreadPool } from '../../snarky.js';
+import { Provable } from '../provable/provable.js';
+import { snarkContext, gatesFromJson } from '../provable/core/provable-context.js';
+import { prettifyStacktrace, prettifyStacktracePromise } from '../util/errors.js';
+import { ProvablePure } from '../provable/types/provable-intf.js';
+
+// external API
+export { ZkFunction, Keypair, Proof, VerificationKey };
+
+type ZkFunctionConfig<P, W extends any[]> = {
+  name: string;
+  publicInputType: ProvablePure<P>;
+  privateInputTypes: { [K in keyof W]: ProvablePure<W[K]> };
+  main: (publicInput: P, ...privateInputs: W) => void;
+};
+
+function ZkFunction<P, W extends any[]>(config: ZkFunctionConfig<P, W>) {
+  const { publicInputType } = config;
+  return {
+    /**
+     * Generates a proving key and a verification key for this circuit(ZkFunction).
+     * @example
+     * ```ts
+     * const keypair = await zkf.generateKeypair();
+     * ```
+     */
+    async generateKeypair() {
+      const main = mainFromCircuitData(config);
+      const publicInputSize = config.publicInputType.sizeInFields();
+      await initializeBindings();
+      return prettifyStacktracePromise(
+        withThreadPool(async () => {
+          let keypair = Snarky.circuit.compile(main, publicInputSize);
+          return new Keypair(keypair);
+        })
+      );
+    },
+
+    /**
+     * Proves a statement using the public input, private input, and the {@link Keypair} of the circuit(ZkFunction).
+     * @example
+     * ```ts
+     * const keypair = await zkf.generateKeypair();
+     * const proof = await zkf.prove([privateInput], publicInput, keypair);
+     * ```
+     */
+    async prove(privateInputs: W, publicInput: P, keypair: Keypair) {
```

Comment:
right now, private inputs seem to be expected in the form of an array - i think it would be good to split them up as well as require the public input as first argument